### PR TITLE
[WIP] Add distributed log protocol/primitive

### PR DESCRIPTION
**Pull request by kuujo**
This PR is an initial implementation of a distributed log protocol and primitive re: #903 

The log protocol is implemented by adding a new core primitive API for logs: `LogClient`, `LogSession`, and `LogRecord`. The implementation of these interfaces is the distributed `log` protocol. As with other protocols, a `LogPartitionGroup`/`LogPartition` is used to configure a distributed log, and a `DistributedLogProtocol` is used to configure primitives to use the distributed log. The `DistributedLog` primitive is the only primitive that is not a state machine that supports the `DistributedLogProtocol`.

All other primitives can use the `DistributedLogProtocol` as well. When the `DistributedLogProtocol` is used on a primitive, the `LogProxySession` implements the `ProxySession` for the primitive and handles writing primitive operations to the distributed log and applying operations to a local `PrimitiveService` when reading from the distributed log. Reads are performed locally in a sequentially consistent manner.

The distributed log protocol itself works much like the primary-backup protocol. It uses primary elections to elect a leader and followers and replicates logs from the leader to the followers. The leader is responsible for all client requests. Client consumers receive distributed log records from the leader over UDP to reduce overhead, and missing log records are resolved by tracking indexes on the client to ensure records are received in sequential order. When the network drops a log record, the client requests records starting at the last received index from the leader.

Logs are persisted to disk on all participating nodes using a `SegmentedJournal`. Nodes support size or time-based compaction of the distributed log, retaining as much information as possible as opposed to e.g. Raft which tends to compact logs sooner.

The `DistributedLog` primitive does not track and store consumer offsets. Instead, users can use other primitives (e.g. `AtomicCounter` or `AtomicCounterMap`) to track consumer offsets, and the `DistributedLog` API provides a mechanism for consumers to start at an arbitrary offset.

Distributed log primitives are not named like other primitives since a `DistributedLog` primitive reads all entries in a distributed log partition group.

```java
DistributedLog<String> log = atomix.<String>logBuilder()
  .withProtocol(DistributedLogProtocol.builder()
    ...
    .build())
  .build();

log.produce("Hello world!");
log.consume(record -> {
  long offset = record.offset();
  long timestamp = record.timestamp();
  String value = record.value();
});
```

### CODE REVIEW

**Comment by kuujo** [comment]
BTW there were a few bugs fixed in this PR relating to primary elections and cluster communication. I will submit them separately as PRs to the 3.0 branch as well.

File: core/src/main/java/io/atomix/core/log/impl/DefaultAsyncDistributedLog.java [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    client.getPartitions().forEach(partition -> {
+      DefaultAsyncDistributedLogPartition<E> logPartition = new DefaultAsyncDistributedLogPartition<>(this, partition, serializer);
+      partitions.put(partition.partitionId().id(), logPartition);
+      sortedPartitions.add(logPartition);
+    });
+  }
+
+  @Override
+  public String name() {
+    return null;
```
Comment (johnou) [inline core/src/main/java/io/atomix/core/log/impl/DefaultAsyncDistributedLog.java]: shouldn't this be assigned?

File: primitive/src/main/java/io/atomix/primitive/session/impl/AbstractSession.java [hunk 2]
```diff
@@ -47,7 +47,7 @@ protected AbstractSession(
     this.sessionId = checkNotNull(sessionId);
     this.primitiveName = checkNotNull(primitiveName);
     this.primitiveType = checkNotNull(primitiveType);
-    this.memberId = checkNotNull(memberId);
+    this.memberId = memberId;
```
Comment (johnou) [inline primitive/src/main/java/io/atomix/primitive/session/impl/AbstractSession.java]: was this an accident?

File: protocols/log/src/main/java/io/atomix/protocols/log/serializer/impl/LogNamespaces.java [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+      .register(ConsumeRequest.class)
+      .register(ConsumeResponse.class)
+      .register(RecordsRequest.class)
+      .register(ResetRequest.class)
+      .register(BackupRequest.class)
+      .register(BackupResponse.class)
+      .register(BackupOperation.class)
+      .register(LogEntry.class)
+      .register(LogRecord.class)
+      .build("PrimaryBackupProtocol");
```
Comment (johnou) [inline protocols/log/src/main/java/io/atomix/protocols/log/serializer/impl/LogNamespaces.java]: shouldn't this read LogProtocol?

File: protocols/log/src/main/java/io/atomix/protocols/log/impl/DistributedLogSession.java [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+  private final ThreadContext threadContext;
+  private final Set<Consumer<PrimitiveState>> stateChangeListeners = Sets.newIdentityHashSet();
+  private final ClusterMembershipEventListener membershipEventListener = this::handleClusterEvent;
+  private final PrimaryElectionEventListener primaryElectionListener = event -> changeReplicas(event.term());
+  private final DistributedLogProducer producer = new DistributedLogProducer();
+  private final DistributedLogConsumer consumer = new DistributedLogConsumer();
+  private final MemberId memberId;
+  private final String subject;
+  private PrimaryTerm term;
+  private volatile PrimitiveState state = PrimitiveState.CONNECTED;
```
Comment (johnou) [inline protocols/log/src/main/java/io/atomix/protocols/log/impl/DistributedLogSession.java]: shouldn't the init state be closed?

File: protocols/log/src/main/java/io/atomix/protocols/log/DistributedLogSessionClient.java [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+    public DistributedLogSessionClient build() {
+      Logger log = ContextualLoggerFactory.getLogger(DistributedLogSessionClient.class, LoggerContext.builder(DistributedLogSessionClient.class)
+          .addValue(clientName)
+          .build());
+
+      // If a ThreadContextFactory was not provided, create one and ensure it's closed when the client is stopped.
+      boolean closeOnStop;
+      ThreadContextFactory threadContextFactory;
+      if (this.threadContextFactory == null) {
+        threadContextFactory = threadModel.factory("backup-client-" + clientName + "-%d", threadPoolSize, log);
```
Comment (kuujo) [inline protocols/log/src/main/java/io/atomix/protocols/log/DistributedLogSessionClient.java]: The thread name needs to be changed

File: primitive/src/main/java/io/atomix/primitive/session/impl/AbstractSession.java (hunk 2 above)
Comment (kuujo) [reply primitive/src/main/java/io/atomix/primitive/session/impl/AbstractSession.java]: No it's intentional. There's no member information with distributed log entries, so the sessions just set it to `null`

File: protocols/log/src/main/java/io/atomix/protocols/log/roles/LogServerRole.java [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+   * @return the role type
+   */
+  public Role role() {
+    return role;
+  }
+
+  /**
+   * Logs a request.
+   */
+  protected final <R extends LogRequest> R logRequest(R request) {
```
[the remark this answers is not in the document]
Comment (kuujo) [reply protocols/log/src/main/java/io/atomix/protocols/log/roles/LogServerRole.java]: Received request, sent response

**Pull request merged**